beacon-chain/sync/initial-sync/blocks_queue.go

```diff
@@ -0,0 +1,413 @@
+					}
+					wg.Done()
+				}()
+
+				skippedBlocks, err := q.parseFetchResponse(q.ctx, response)
+				if err != nil {
+					q.state.scheduler.incrementCounter(failedBlock, response.count)
+					return
+				}
+				q.state.scheduler.incrementCounter(skippedBlock, skippedBlocks)
+			}()
+		case <-q.pendingFetchedBlocks:
+			wg.Add(1)
+			go func() {
+				defer wg.Done()
+				if err := q.sendFetchedBlocks(q.ctx); err != nil {
+					log.WithError(err).Debug("Error sending received blocks")
+				}
+			}()
+		}
+	}
+}
+
+// scheduleFetchRequests enqueues block fetch requests to block fetcher.
+func (q *blocksQueue) scheduleFetchRequests(ctx context.Context) error {
+	q.state.scheduler.Lock()
+	defer q.state.scheduler.Unlock()
+
+	if ctx.Err() != nil {
+		return ctx.Err()
+	}
+
+	s := q.state.scheduler
+	blocks := q.state.scheduler.requestedBlocks
+
+	func() {
+		resetStateCounters := func() {
+			for i := 0; i < blockStateLen; i++ {
+				blocks[blockState(i)] = 0
+			}
+			s.currentSlot = q.headFetcher.HeadSlot()
+		}
+
+		// Update state's current slot pointer.
+		count := blocks[pendingBlock] + blocks[skippedBlock] + blocks[failedBlock] + blocks[validBlock]
+		if count == 0 {
+			s.currentSlot = q.headFetcher.HeadSlot()
+			return
+		}
+
+		// Too many failures (blocks that can't be processed at this time).
+		if blocks[failedBlock] >= s.blockBatchSize/2 {
+			s.blockBatchSize *= 2
+			resetStateCounters()
+			return
+		}
+
+		// Given enough valid blocks, we can set back block batch size.
+		if blocks[validBlock] >= blockBatchSize && s.blockBatchSize != blockBatchSize {
+			blocks[skippedBlock], blocks[validBlock] = blocks[skippedBlock]+blocks[validBlock], 0
+			s.blockBatchSize = blockBatchSize
+		}
+
+		// Too many items in scheduler, time to update current slot to point to current head's slot.
+		count = blocks[pendingBlock] + blocks[skippedBlock] + blocks[failedBlock] + blocks[validBlock]
+		if count >= queueMaxCachedBlocks {
+			s.blockBatchSize = blockBatchSize
+			resetStateCounters()
+			return
+		}
+
+		// All blocks processed, no pending blocks.
+		count = blocks[skippedBlock] + blocks[failedBlock] + blocks[validBlock]
+		if count > 0 && blocks[pendingBlock] == 0 {
+			s.blockBatchSize = blockBatchSize
+			resetStateCounters()
+			return
+		}
+	}()
+
+	offset := blocks[pendingBlock] + blocks[skippedBlock] + blocks[failedBlock] + blocks[validBlock]
+	start := q.state.scheduler.currentSlot + offset + 1
+	count := mathutil.Min(q.state.scheduler.blockBatchSize, q.highestExpectedSlot-start+1)
+	if count <= 0 {
+		return errStartSlotIsTooHigh
+	}
+
+	ctx, _ = context.WithTimeout(ctx, queueFetchRequestTimeout)
+	if err := q.blocksFetcher.scheduleRequest(ctx, start, count); err != nil {
+		return err
+	}
+	q.state.scheduler.requestedBlocks[pendingBlock] += count
+
+	return nil
+}
+
+// parseFetchResponse processes incoming responses.
+func (q *blocksQueue) parseFetchResponse(ctx context.Context, response *fetchRequestResponse) (uint64, error) {
```

Comment:
can you add a span here?

Reply:
discussed offline, there's spans in fetcher for every functions that might take some time, this is likely not needed & to avoid potential deadlocks